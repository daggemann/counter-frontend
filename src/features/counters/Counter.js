import axios from 'axios';

const API_URL = process.env.REACT_APP_API_URL + '/counters';

export class Counter {

    constructor(id, title, count) {
        this._id = id;
        this._title = title;
        this._count = count;
    }

    get id() {
        return this._id;
    }

    set id(id) {
        this._id = id;
    }

    get title() {
        return this._title;
    }

    set title(title) {
        this._title = title;
    }

    get count() {
        return this._count;
    }

    set count(count) {
        this._count = count;
    }

    increment() {
        return new Promise( (resolve, reject) => {
            axios.post(API_URL + `/${this._id}/increment`, {})
                .then(res => {
                    const counters = res.data.map( (counter) => {
                        return new Counter(counter.id, counter.title, counter.count);
                    });
                    resolve(counters);
                })
                .catch(err => {
                    reject(err)
                })
        })
    }

    delete() {
        return new Promise( (resolve, reject) => {
            axios.delete(API_URL + `/${this._id}`)
                .then(res => {
                    const counters = res.data.map( (counter) => {
                        return new Counter(counter.id, counter.title, counter.count);
                    });
                    resolve(counters);
                })
                .catch(err => {
                    reject(err)
                })
        })
    }

    decrement() {
        return new Promise( (resolve, reject) => {
            axios.post(API_URL + `/${this._id}/decrement`, {})
                .then(res => {
                    const counters = res.data.map( (counter) => {
                        return new Counter(counter.id, counter.title, counter.count);
                    });
                    resolve(counters);
                })
                .catch(err => {
                    reject(err)
                })
        })
    }
}


export function totalCount(counters) {
    let totalCount = 0;
    counters.forEach((counter) => {
        totalCount += counter.count;
    });
    return totalCount;
}

export function getAll() {
    return new Promise((resolve, reject) => {
        axios.get(API_URL)
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })

    })
}

export function add(title) {
    return new Promise( (resolve, reject) => {
        axios.post(API_URL, {title: title})
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}

export function incrementCounter(id) {
    return new Promise( (resolve, reject) => {
        axios.post(API_URL + `/${id}/increment`, {})
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}

export function decrementCounter(id) {
    return new Promise( (resolve, reject) => {
        axios.post(API_URL + `/${id}/decrement`, {})
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}

export function deleteCounter(id) {
    return new Promise( (resolve, reject) => {
        axios.delete(API_URL + `/${id}`)
            .then(res => {
                const counters = res.data.map( (counter) => {
                    return new Counter(counter.id, counter.title, counter.count);
                });
                resolve(counters);
            })
            .catch(err => {
                reject(err)
            })
    })
}
