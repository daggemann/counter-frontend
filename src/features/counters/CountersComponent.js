import React from 'react';
import AddCounterContainer from './AddCounterContainer';
import TotalCountComponent from './TotalCountComponent';
import CounterComponent from './CounterComponent';
import {FlexCol, HorizontalLine} from "../../commons/styled-components";

export default function CountersComponent(props) {
    return (
        <FlexCol
            margin='20px'
            padding='5px'
            width='300px'
        >
            <AddCounterContainer
                onClick={(title) => props.addCounter(title)}
            />
            <HorizontalLine
                margin='20px 0px 10px 0px'
                borderWidth='0px 0px 1px 0px'
                borderStyle='dashed'
                borderColor='gray'
            />

            {props.counters.map((counter) => {
                return <CounterComponent
                    key={counter.id}
                    counter={counter}
                    decr={(i) => props.decrementCounter(i)}
                    incr={(i) => props.incrementCounter(i)}
                    delete={(i) => props.deleteCounter(i)}
                />
            })}

            <HorizontalLine
                margin='20px 0px 0px 0px'
                borderWidth='0px 0px 1px 0px'
                borderStyle='dashed'
                borderColor='gray'
            />
            <TotalCountComponent total={props.totalCount}/>
        </FlexCol>
    )
}

