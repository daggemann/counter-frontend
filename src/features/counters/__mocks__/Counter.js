export class Counter {
    constructor(id, title, count) {
        this.id = id;
        this.title = title;
        this.count = count;
    }

    delete() {
        delete counters[this.id]
    }
}

let counters = [
    new Counter(1, "test", 1),
    new Counter(2, "test", 1),
    new Counter(3, "test", 1),
];


export function getAll() {
    return new Promise((resolve, reject) => {
        (true) ? setTimeout(resolve(counters.slice()), 3) : reject('never gonna happen')
    });
}

export function add(title) {
    counters.push(new Counter(4, "test", 1));
    return new Promise((resolve, reject) => {
        (true) ? setTimeout(resolve(counters.slice()), 3) : reject('never gonna happen')
    });
}

export function totalCount(counters) {
    let totalCount = 0;
    counters.forEach((counter) => {
        totalCount += counter.count;
    });
    return totalCount;
}

