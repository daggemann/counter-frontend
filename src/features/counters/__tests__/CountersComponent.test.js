import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import {Counter} from "../Counter";
import CountersComponent from "../CountersComponent";


test('render counters with no counters', () => {
    const counters = [];
    const tree = renderer.create(<CountersComponent counters={counters} totalCount={0}/>)
    expect(tree.toJSON()).toMatchSnapshot();
});

test('render counters with some counters that have no count', () => {
    const counters = [
        new Counter(1, "test", 0),
        new Counter(2, "test", 0),
        new Counter(3, "test", 0),
    ];
    const tree = renderer.create(<CountersComponent counters={counters} totalCount={0}/>)
    expect(tree.toJSON()).toMatchSnapshot();
});

test('render counters with some counters that have some count', () => {
    const counters = [
        new Counter(1, "test", 1),
        new Counter(2, "test", -2),
        new Counter(3, "test", 5),
    ];
    const tree = renderer.create(<CountersComponent counters={counters} totalCount={4}/>)
    expect(tree.toJSON()).toMatchSnapshot();
});

test('ensure that correct amount of CounterComponent are added', () => {
    const counters = [
        new Counter(1, "test", 1),
        new Counter(2, "test", -2),
        new Counter(3, "test", 5),
    ];
    const countersComponent = shallow(<CountersComponent counters={counters} totalCount={4}/>)
    const counterComponents = countersComponent.find('CounterComponent');
    expect(counterComponents.length).toEqual(3)
});

test('ensure that NewCounterContainer is rendered', () => {
    const counters = [];
    const countersComponent = shallow(<CountersComponent counters={counters} totalCount={0}/>)
    const addCounterComponent = countersComponent.find('NewCounterContainer');
    expect(addCounterComponent.length).toEqual(1)
});

test('ensure that TotalCountComponent is rendered', () => {
    const counters = [];
    const countersComponent = shallow(<CountersComponent counters={counters} totalCount={0}/>)
    const totalCountComponent = countersComponent.find('TotalCountComponent');
    expect(totalCountComponent.length).toEqual(1)
});
