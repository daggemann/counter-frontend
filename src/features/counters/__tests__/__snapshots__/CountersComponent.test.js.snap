// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render counters with no counters 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  height: 30px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin: 10px 0px 0px 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  margin: 20px;
  padding: 5px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  margin: 20px 0px 10px 0px;
  border-width: 0px 0px 1px 0px;
  border-style: dashed;
  border-color: gray;
}

.c5 {
  margin: 20px 0px 0px 0px;
  border-width: 0px 0px 1px 0px;
  border-style: dashed;
  border-color: gray;
}

.c2 {
  border-width: 0px 0px 1px 0px;
  border-style: solid;
  border-color: gray;
}

.c3 {
  border: 1px solid gray;
  border-radius: 1em;
  width: 100px;
}

.c3:hover {
  background-color: #2ECC40;
}

.c7 {
  margin-left: 10px;
}

<div
  className="c0"
  width="300px"
>
  <div
    className="c1"
    height="30px"
    width="300px"
  >
    <input
      className="c2"
      onChange={[Function]}
      placeholder="Counter title"
      size={15}
      type="text"
      value=""
    />
    <button
      className="c3"
      onClick={[Function]}
      width="100px"
    >
      Add
    </button>
  </div>
  <div
    className="c4"
  />
  <div
    className="c5"
  />
  <div
    className="c6"
  >
    Total:
    <span
      className="c7"
    >
      0
    </span>
  </div>
</div>
`;

exports[`render counters with some counters that have no count 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  height: 30px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin: 10px 0px 0px 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 30px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 5px 0px 0px 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  margin: 20px;
  padding: 5px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  margin: 20px 0px 10px 0px;
  border-width: 0px 0px 1px 0px;
  border-style: dashed;
  border-color: gray;
}

.c13 {
  margin: 20px 0px 0px 0px;
  border-width: 0px 0px 1px 0px;
  border-style: dashed;
  border-color: gray;
}

.c2 {
  border-width: 0px 0px 1px 0px;
  border-style: solid;
  border-color: gray;
}

.c3 {
  border: 1px solid gray;
  border-radius: 1em;
  width: 100px;
}

.c3:hover {
  background-color: #2ECC40;
}

.c15 {
  margin-left: 10px;
}

.c7 {
  border: 0px;
  border-radius: 1em 0em 0em 1em;
  width: 30px;
  background-color: #FF4136;
  margin: 0px 2px 0px 0px;
}

.c7:hover {
  margin-right: 10px;
}

.c8 {
  margin-top: 6px;
  text-align: center;
}

.c10 {
  border: 0px;
  border-radius: 1em 0em 0em 1em;
  width: 20px;
}

.c10:hover {
  background-color: #7FDBFF;
}

.c11 {
  text-align: center;
  margin-top: 6px;
  width: 50px;
}

.c12 {
  border: 0px;
  border-radius: 0em 1em 1em 0em;
  width: 20px;
}

.c12:hover {
  background-color: #7FDBFF;
}

<div
  className="c0"
  width="300px"
>
  <div
    className="c1"
    height="30px"
    width="300px"
  >
    <input
      className="c2"
      onChange={[Function]}
      placeholder="Counter title"
      size={15}
      type="text"
      value=""
    />
    <button
      className="c3"
      onClick={[Function]}
      width="100px"
    >
      Add
    </button>
  </div>
  <div
    className="c4"
  />
  <div
    className="c5"
    height="30px"
  >
    <div
      className="c6"
    >
      <button
        className="c7"
        onClick={[Function]}
        width="30px"
      >
        x
      </button>
      <span
        className="c8"
      >
        test
      </span>
    </div>
    <div
      className="c9"
      width="100px"
    >
      <button
        className="c10"
        onClick={[Function]}
        width="20px"
      >
        -
      </button>
      <span
        className="c11"
      >
        0
      </span>
      <button
        className="c12"
        onClick={[Function]}
        width="20px"
      >
        +
      </button>
    </div>
  </div>
  <div
    className="c5"
    height="30px"
  >
    <div
      className="c6"
    >
      <button
        className="c7"
        onClick={[Function]}
        width="30px"
      >
        x
      </button>
      <span
        className="c8"
      >
        test
      </span>
    </div>
    <div
      className="c9"
      width="100px"
    >
      <button
        className="c10"
        onClick={[Function]}
        width="20px"
      >
        -
      </button>
      <span
        className="c11"
      >
        0
      </span>
      <button
        className="c12"
        onClick={[Function]}
        width="20px"
      >
        +
      </button>
    </div>
  </div>
  <div
    className="c5"
    height="30px"
  >
    <div
      className="c6"
    >
      <button
        className="c7"
        onClick={[Function]}
        width="30px"
      >
        x
      </button>
      <span
        className="c8"
      >
        test
      </span>
    </div>
    <div
      className="c9"
      width="100px"
    >
      <button
        className="c10"
        onClick={[Function]}
        width="20px"
      >
        -
      </button>
      <span
        className="c11"
      >
        0
      </span>
      <button
        className="c12"
        onClick={[Function]}
        width="20px"
      >
        +
      </button>
    </div>
  </div>
  <div
    className="c13"
  />
  <div
    className="c14"
  >
    Total:
    <span
      className="c15"
    >
      0
    </span>
  </div>
</div>
`;

exports[`render counters with some counters that have some count 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  height: 30px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin: 10px 0px 0px 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 30px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 5px 0px 0px 0px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 300px;
  margin: 20px;
  padding: 5px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  margin: 20px 0px 10px 0px;
  border-width: 0px 0px 1px 0px;
  border-style: dashed;
  border-color: gray;
}

.c13 {
  margin: 20px 0px 0px 0px;
  border-width: 0px 0px 1px 0px;
  border-style: dashed;
  border-color: gray;
}

.c2 {
  border-width: 0px 0px 1px 0px;
  border-style: solid;
  border-color: gray;
}

.c3 {
  border: 1px solid gray;
  border-radius: 1em;
  width: 100px;
}

.c3:hover {
  background-color: #2ECC40;
}

.c15 {
  margin-left: 10px;
}

.c7 {
  border: 0px;
  border-radius: 1em 0em 0em 1em;
  width: 30px;
  background-color: #FF4136;
  margin: 0px 2px 0px 0px;
}

.c7:hover {
  margin-right: 10px;
}

.c8 {
  margin-top: 6px;
  text-align: center;
}

.c10 {
  border: 0px;
  border-radius: 1em 0em 0em 1em;
  width: 20px;
}

.c10:hover {
  background-color: #7FDBFF;
}

.c11 {
  text-align: center;
  margin-top: 6px;
  width: 50px;
}

.c12 {
  border: 0px;
  border-radius: 0em 1em 1em 0em;
  width: 20px;
}

.c12:hover {
  background-color: #7FDBFF;
}

<div
  className="c0"
  width="300px"
>
  <div
    className="c1"
    height="30px"
    width="300px"
  >
    <input
      className="c2"
      onChange={[Function]}
      placeholder="Counter title"
      size={15}
      type="text"
      value=""
    />
    <button
      className="c3"
      onClick={[Function]}
      width="100px"
    >
      Add
    </button>
  </div>
  <div
    className="c4"
  />
  <div
    className="c5"
    height="30px"
  >
    <div
      className="c6"
    >
      <button
        className="c7"
        onClick={[Function]}
        width="30px"
      >
        x
      </button>
      <span
        className="c8"
      >
        test
      </span>
    </div>
    <div
      className="c9"
      width="100px"
    >
      <button
        className="c10"
        onClick={[Function]}
        width="20px"
      >
        -
      </button>
      <span
        className="c11"
      >
        1
      </span>
      <button
        className="c12"
        onClick={[Function]}
        width="20px"
      >
        +
      </button>
    </div>
  </div>
  <div
    className="c5"
    height="30px"
  >
    <div
      className="c6"
    >
      <button
        className="c7"
        onClick={[Function]}
        width="30px"
      >
        x
      </button>
      <span
        className="c8"
      >
        test
      </span>
    </div>
    <div
      className="c9"
      width="100px"
    >
      <button
        className="c10"
        onClick={[Function]}
        width="20px"
      >
        -
      </button>
      <span
        className="c11"
      >
        -2
      </span>
      <button
        className="c12"
        onClick={[Function]}
        width="20px"
      >
        +
      </button>
    </div>
  </div>
  <div
    className="c5"
    height="30px"
  >
    <div
      className="c6"
    >
      <button
        className="c7"
        onClick={[Function]}
        width="30px"
      >
        x
      </button>
      <span
        className="c8"
      >
        test
      </span>
    </div>
    <div
      className="c9"
      width="100px"
    >
      <button
        className="c10"
        onClick={[Function]}
        width="20px"
      >
        -
      </button>
      <span
        className="c11"
      >
        5
      </span>
      <button
        className="c12"
        onClick={[Function]}
        width="20px"
      >
        +
      </button>
    </div>
  </div>
  <div
    className="c13"
  />
  <div
    className="c14"
  >
    Total:
    <span
      className="c15"
    >
      4
    </span>
  </div>
</div>
`;
